cmds/dhclient/dhcp6client/dhcp6client_test.go

```diff
@@ -0,0 +1,329 @@
+)
+
+func TestSolicitMessage(t *testing.T) {
+	p, _ := newSolicitPacket(&mac)
+	pb, _ := p.MarshalBinary()
+
+	r := &testMessage{
+		addr: &net.UDPAddr{
+			IP: net.ParseIP(v6addr),
+		},
+	}
+	r.b.Write(pb)
+
+	reply, err := serve(r)
+	if err != nil {
+		t.Fatalf("%v\n", err)
+	}
+	t.Logf("%v\n", reply)
+	// TODO: return reply and see if as expected
+}
+
+type PacketConn dhcp6.PacketConn
+
+type testMessage struct {
+	b    bytes.Buffer
+	cm   *ipv6.ControlMessage
+	addr net.Addr
+}
+
+/**
+* recordIPv6PacketConn
+ */
+type recordIPv6PacketConn struct {
+	closed bool
+	joined []net.Addr
+	left   []net.Addr
+	flags  map[ipv6.ControlFlags]bool
+}
+
+func (c *recordIPv6PacketConn) Close() error {
+	c.closed = true
+	return nil
+}
+
+func (c *recordIPv6PacketConn) JoinGroup(iface *net.Interface, group net.Addr) error {
+	c.joined = append(c.joined, group)
+	return nil
+}
+
+func (c *recordIPv6PacketConn) LeaveGroup(iface *net.Interface, group net.Addr) error {
+	c.left = append(c.left, group)
+	return nil
+}
+
+func (c *recordIPv6PacketConn) SetControlMessage(cf ipv6.ControlFlags, on bool) error {
+	c.flags[cf] = on
+	return nil
+}
+
+/**
+* testPacketConn
+ */
+type testPacketConn struct {
+	r *testMessage
+	w *testMessage
+
+	*recordIPv6PacketConn
+}
+
+func (c *testPacketConn) ReadFrom(b []byte) (int, *ipv6.ControlMessage, net.Addr, error) {
+	n, err := c.r.b.Read(b)
+	cm := c.r.cm
+	src := c.r.addr
+
+	return n, cm, src, err
+}
+
+func (c *testPacketConn) WriteTo(b []byte, cm *ipv6.ControlMessage, dst net.Addr) (int, error) {
+	n, err := c.w.b.Write(b)
+	c.w.cm = cm
+	c.w.addr = dst
+
+	return n, err
+}
+
+/**
+* testPacketConn
+ */
+type oneReadPacketConn struct {
+	PacketConn
+
+	err    error
+	txDone bool
+
+	readDoneChan  chan struct{}
+	writeDoneChan chan struct{}
+}
+
+func (c *oneReadPacketConn) ReadFrom(b []byte) (int, *ipv6.ControlMessage, net.Addr, error) {
+	if c.txDone {
+		return 0, nil, nil, errClosing
+	}
+	c.txDone = true
+	n, cm, addr, err := c.PacketConn.ReadFrom(b)
+	close(c.readDoneChan)
+	log.Printf("read from: %v, %v, %v, %v\n", n, cm, addr, err)
+	return n, cm, addr, err
+}
+
+func (c *oneReadPacketConn) WriteTo(b []byte, cm *ipv6.ControlMessage, addr net.Addr) (int, error) {
+	n, err := c.PacketConn.WriteTo(b, cm, addr)
+	close(c.writeDoneChan)
+	return n, err
+}
+
+func serve(r *testMessage) (*dhcp6.Packet, error) {
+	s := &dhcp6.Server{}
+	s.Iface = &net.Interface{
+		Name:  "foo",
+		Index: 0,
+	}
+	s.Handler = &Handler{
+		ip:      net.ParseIP("::0"),
+		handler: handle,
+	}
+
+	tc := &testPacketConn{
+		r: r,
+		w: &testMessage{},
+
+		recordIPv6PacketConn: &recordIPv6PacketConn{
+			closed: false,
+			joined: make([]net.Addr, 0),
+			left:   make([]net.Addr, 0),
+			flags:  make(map[ipv6.ControlFlags]bool),
+		},
+	}
+
+	c := &oneReadPacketConn{
+		PacketConn: tc,
+		txDone:     false,
+
+		readDoneChan:  make(chan struct{}, 0),
+		writeDoneChan: make(chan struct{}, 0),
+	}
+
+	err := s.Serve(c)
+	<-c.readDoneChan
```

Comment:
Agree. 

So the usage of this channel is: normally the server listens with a infinite loop, whereas in test we only need server to listen once. So after the server reads from the connection for the first time, the channel will be closed, and the connection returns errClosing. Please see oneReadPackConn.ReadFrom() as reference.

I get this idea basically from https://github.com/mdlayher/dhcp6/blob/0570f4289fd3465c31d7f27e5ba36ad2800d60e2/server_test.go#L448